### PR TITLE
Finance 121

### PR DESCRIPTION
**Pull request by jgheerardyn**

### CODE REVIEW

File: foundations/src/main/java/io/yields/math/framework/definition/TestDefinitions.java [hunk 1]
```diff
@@ -4,7 +4,7 @@
 import java.util.List;
 
 /**
- * Copyright (C) 07/08/15 - REstore NV
+ * Copyright (C) 07/08/15 - Yields NV
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/TestDefinitions.java]: oeps, did it again, grrrrrr. Thx for fixing.

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobDefinition.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+/**
+ * Describes which test(s) need to be scheduled and when
+ */
+public class TestJobDefinition {
+
+    private List<String> testDefinitionNames;
+    private String author;
+    private TestJobSchedule schedule;
+
+    public TestJobDefinition(List<String> testDefinitionNames, String author, TestJobSchedule schedule) {
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobDefinition.java]: don't you need a 'project' or 'application', i.e. linking to the job on Jenkins which needs to be kicked with the given test definition names?

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobResult.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ * represents the result of running a certain test job
+ */
+public class TestJobResult {
+
+    private String testDefinitionName;
+    private LocalDateTime executionTime;
+    private TestJobResultType resultType;
+    private String errorMessage;
+
+    public static TestJobResult success(String testDefinitionName, LocalDateTime executionTime) {
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobResult.java]: what is execution time: start or end? why not have both?

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobResultType.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.yields.math.framework.definition;
+
+public enum TestJobResultType {
+    SUCCEEDED, FAILED;
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobResultType.java]: If you look at the build status in Jenkins, there's a difference between FAILED (i.e. job actually ran, but 1 or more tests failed) and ERROR (compilation error, or eg. jar missing).

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobScheduleType.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.yields.math.framework.definition;
+
+/**
+ * indicates when tests should be indicated
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobScheduleType.java]: indicated?

File: foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+package io.yields.math.framework.definition.json;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.StringWriter;
+
+import io.yields.math.framework.definition.TestJobDefinitions;
+
+public class TestJobDefinitionJsonExporter {
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java]: plural: TestJobDefinitionsJsonExporter

File: foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    private static ObjectMapper mapper = new ObjectMapper();
+
+    public static TestJobDefinitions deserialize(String content) {
+        try {
+            return mapper.readValue(content, TestJobDefinitions.class);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("Could not deserialize test job", e);
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java]: be consistent: test job definitions

File: foundations/src/main/java/io/yields/math/framework/kpi/StandardResult.java [hunk 8]
```diff
@@ -19,7 +19,7 @@
 import java.util.List;
 
 /**
- * Copyright (C) 11/03/15 - REstore NV
+ * Copyright (C) 11/03/15 - Yields NV
```
Comment (rombautt) [inline foundations/src/main/java/io/yields/math/framework/kpi/StandardResult.java]: aaaaargh

File: foundations/src/test/java/io/yields/math/framework/definition/TestJobScheduleTest.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+
+package io.yields.math.framework.definition;
+
+import org.junit.Test;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+public class TestJobScheduleTest {
```
Comment (rombautt) [inline foundations/src/test/java/io/yields/math/framework/definition/TestJobScheduleTest.java]: missing tests for the other types of test job schedules

File: foundations/src/test/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporterTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        TestJobDefinitions deserialized = TestJobDefinitionJsonExporter.deserialize(json);
+        assertThat(deserialized.getTestJobDefinitions()).hasSize(2);
+        TestJobDefinition oneOffDeser = deserialized.getTestJobDefinitions().get(0);
+        TestJobDefinition periodicalDeser = deserialized.getTestJobDefinitions().get(1);
+        assertEqual(oneOff, oneOffDeser);
+        assertEqual(periodical, periodicalDeser);
+    }
+
+    private void assertEqual(TestJobDefinition original, TestJobDefinition deserialized) {
+        assertThat(original.getAuthor()).isEqualTo(deserialized.getAuthor());
```
Comment (rombautt) [inline foundations/src/test/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporterTest.java]: why not implement a decent equals on TestJobDefinition, and use assertThat().containsExactly

**Comment by rombautt** [comment]
reviewed

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobDefinition.java (hunk 2 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobDefinition.java]: done

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobResult.java (hunk 3 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobResult.java]: added start and end

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobResultType.java (hunk 4 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobResultType.java]: added error

File: foundations/src/main/java/io/yields/math/framework/definition/TestJobScheduleType.java (hunk 5 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/TestJobScheduleType.java]: lol

File: foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java (hunk 6 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java]: ok - in that case will apply plural to TestDefinitionJsonExplorer as well...

File: foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java (hunk 7 above)
Comment (jgheerardyn) [inline foundations/src/main/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporter.java]: changed

File: foundations/src/test/java/io/yields/math/framework/definition/TestJobScheduleTest.java (hunk 9 above)
Comment (jgheerardyn) [inline foundations/src/test/java/io/yields/math/framework/definition/TestJobScheduleTest.java]: added - but they look quite trivial...

File: foundations/src/test/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporterTest.java (hunk 10 above)
Comment (jgheerardyn) [inline foundations/src/test/java/io/yields/math/framework/definition/json/TestJobDefinitionJsonExporterTest.java]: done

**Pull request merged**